In [1]:
!pip install scikit-learn==1.2.2
!pip install scikeras --upgrade

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.6/9.6 MB 22.5 MB/s eta 0:00:00
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.3.2
    Uninstalling scikit-learn-1.3.2:
      Successfully uninstalled scikit-learn-1.3.2
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.4/13.4 MB 41.6 MB/s eta 0:00:00
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.2.2
    Uninstalling scikit-learn-1.2.2:
      Successfully uninstalled scikit-learn-1.2.2


In [2]:
# Import our dependencies
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import classification_report, confusion_matrix
import pandas as pd
import tensorflow as tf

#  Import and read the diabetes_data.csv.
import pandas as pd
diabetes_df = pd.read_csv('https://raw.githubusercontent.com/ComfyKoala/diabetes-classification/main/Maribel/diabetes_data_upload.csv')


In [3]:
#Dropping strongest features "Gender", "Polyuria", "Polydipsia"
diabetes_df=diabetes_df.drop(columns=["Gender", "Polyuria", "Polydipsia"])
diabetes_df.head()

Age sudden weight loss weakness Polyphagia Genital thrush visual blurring  \
0   40                 No      Yes         No             No              No   
1   58                 No      Yes         No             No             Yes   
2   41                 No      Yes        Yes             No              No   
3   45                Yes      Yes        Yes            Yes              No   
4   60                Yes      Yes        Yes             No             Yes   

  Itching Irritability delayed healing partial paresis muscle stiffness  \
0     Yes           No             Yes              No              Yes   
1      No           No              No             Yes               No   
2     Yes           No             Yes              No              Yes   
3     Yes           No             Yes              No               No   
4     Yes          Yes             Yes             Yes              Yes   

  Alopecia Obesity     class  
0      Yes     Yes  Positive  
1      Yes      No  Positive  
2      Yes      No  Positive  
3       No      No  Positive  
4      Yes     Yes  Positive

In [4]:
#change all "yes" to 1 and all "no" to 0
diabetes_df = diabetes_df.replace({"Yes": 1, "No": 0})

#change "female" to 1 and "male" to 0
diabetes_df = diabetes_df.replace({"Female": 1, "Male": 0})

In [5]:
# Get the target variables
target=diabetes_df["class"]
target_names = ["negative", "positive"]

In [6]:
#Get the features
data = diabetes_df.drop("class", axis=1)
feature_names = data.columns
data.head()

Age  sudden weight loss  weakness  Polyphagia  Genital thrush  \
0   40                   0         1           0               0   
1   58                   0         1           0               0   
2   41                   0         1           1               0   
3   45                   1         1           1               1   
4   60                   1         1           1               0   

   visual blurring  Itching  Irritability  delayed healing  partial paresis  \
0                0        1             0                1                0   
1                1        0             0                0                1   
2                0        1             0                1                0   
3                0        1             0                1                0   
4                1        1             1                1                1   

   muscle stiffness  Alopecia  Obesity  
0                 1         1        1  
1                 0         1        0  
2                 1         1        0  
3                 0         0        0  
4                 1         1        1

In [7]:
# Split data into training and testing
X_train, X_test, y_train, y_test = train_test_split(data, target, random_state=42)

## Support Vector Machines

In [8]:
# Support vector machine linear classifier
model = SVC(kernel='linear')
model.fit(X_train, y_train)

SVC(kernel='linear')

In [9]:
# Model Accuracy
print('Test Acc: %.3f' % model.score(X_test, y_test))

Test Acc: 0.769


In [10]:
# Calculate the classification report
from sklearn.metrics import classification_report
predictions = model.predict(X_test)
print(classification_report(y_test, predictions,
                            target_names=target_names))

              precision    recall  f1-score   support

    negative       0.68      0.65      0.67        46
    positive       0.81      0.83      0.82        84

    accuracy                           0.77       130
   macro avg       0.75      0.74      0.75       130
weighted avg       0.77      0.77      0.77       130



In [11]:
#save model to reusable variable
svm_model_77 = model

Removed 3 strongest columns and accuracy went from 92.3% in file of "diabetes1updated" to 76.9% above

# **Identify Weak and Strong Features**

In [12]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [13]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
n_features = X_train.shape[1]

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=100, activation="relu", input_dim=n_features))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=75, activation="sigmoid"))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=50, activation="sigmoid"))

# Fourth hidden layer
nn.add(tf.keras.layers.Dense(units=25, activation="sigmoid"))

# Fifth hidden layer
nn.add(tf.keras.layers.Dense(units=10, activation="sigmoid"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 100)                 │           1,400 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 75)                  │           7,575 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 50)                  │           3,800 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_3 (Dense)                      │ (None, 25)                  │           1,275 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_4 (Dense)                      │ (None, 10)                  │             260 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_5 (Dense)                      │ (None, 1)                   │              11 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 14,321 (55.94 KB)

 Trainable params: 14,321 (55.94 KB)

 Non-trainable params: 0 (0.00 B)

In [14]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [15]:

from scikeras.wrappers import KerasClassifier
from sklearn.inspection import permutation_importance

# Ensure that nn is defined as your Keras model
# wrapped_nn = KerasClassifier(model=nn, epochs=20, batch_size=32, verbose=0)

# Define the wrapped model (assuming nn is your Keras Sequential model)
wrapped_nn = KerasClassifier(model=nn, epochs=20, batch_size=32, verbose=0)

# Train the wrapped model
wrapped_nn.fit(X_train_scaled, y_train)

# Compute permutation importance
result = permutation_importance(wrapped_nn, X_test_scaled, y_test, n_repeats=10, random_state=10)

# Get feature importances
importances = result.importances_mean

# Ensure X_test_scaled is a DataFrame or convert it
if not isinstance(X_test_scaled, pd.DataFrame):
    X_test_scaled = pd.DataFrame(X_test_scaled, columns=X_test.columns)

# Sort the features by importance in ascending order (weakest to strongest)
sorted_indices = np.argsort(importances)

# Get the top 20 weakest features
top_20_weakest_indices = sorted_indices[:20]
top_20_weakest_features = X_test_scaled.columns[top_20_weakest_indices]
top_20_weakest_importances = importances[top_20_weakest_indices]

# Display the weakest features and their importances
for feature, importance in zip(top_20_weakest_features, top_20_weakest_importances):
    print(f"Weakest feature: {feature} with importance {importance}")

Weakest feature: Age with importance 0.01692307692307694
Weakest feature: Obesity with importance 0.016923076923076975
Weakest feature: weakness with importance 0.023076923076923096
Weakest feature: Itching with importance 0.02384615384615386
Weakest feature: Genital thrush with importance 0.02384615384615387
Weakest feature: muscle stiffness with importance 0.0323076923076923
Weakest feature: Polyphagia with importance 0.03615384615384617
Weakest feature: sudden weight loss with importance 0.04076923076923078
Weakest feature: Irritability with importance 0.04384615384615388
Weakest feature: delayed healing with importance 0.05846153846153847
Weakest feature: partial paresis with importance 0.058461538461538474
Weakest feature: visual blurring with importance 0.07153846153846154
Weakest feature: Alopecia with importance 0.07153846153846156


In [16]:
# Wrap your Keras model
wrapped_nn = KerasClassifier(model=nn, epochs=20, batch_size=32, verbose=0)

# Train the wrapped model
wrapped_nn.fit(X_train_scaled, y_train)

# Compute permutation importance
result = permutation_importance(wrapped_nn, X_test_scaled, y_test, n_repeats=10, random_state=10)

# Get feature importances
importances = result.importances_mean

# Sort the features by importance in descending order (strongest to weakest)
sorted_indices = np.argsort(importances)[::-1]

# Get the top 20 strongest features
top_20_strongest_indices = sorted_indices[:20]
top_20_strongest_features = X_test.columns[top_20_strongest_indices]
top_20_strongest_importances = importances[top_20_strongest_indices]

# Display the strongest features and their importances
for feature, importance in zip(top_20_strongest_features, top_20_strongest_importances):
    print(f"Strongest feature: {feature} with importance {importance}")

Strongest feature: visual blurring with importance 0.07384615384615388
Strongest feature: Alopecia with importance 0.06230769230769233
Strongest feature: partial paresis with importance 0.052307692307692354
Strongest feature: sudden weight loss with importance 0.04538461538461543
Strongest feature: Irritability with importance 0.04461538461538468
Strongest feature: Polyphagia with importance 0.03923076923076929
Strongest feature: delayed healing with importance 0.028461538461538503
Strongest feature: muscle stiffness with importance 0.027692307692307728
Strongest feature: weakness with importance 0.02461538461538466
Strongest feature: Genital thrush with importance 0.023076923076923127
Strongest feature: Itching with importance 0.01615384615384622
Strongest feature: Obesity with importance 0.00923076923076931
Strongest feature: Age with importance 0.006153846153846176
